sbi/neural_nets/embedding_nets.py

```diff
@@ -31,3 +39,155 @@ def forward(self, x: Tensor) -> Tensor:
+        Returns:
+            Network output (batch_size, output_dim).
+        """
+        x = self.conv_subnet(x)
+        x = torch.flatten(x, 1)  # flatten all dimensions except batch
+        embedding = self.fc_subnet(x)
+
+        return embedding
+
+
+class FCEmbeddingPermutationInv(nn.Module):
+    """Permutation invariant embedding network.
+
+    Takes as input a tensor with (batch, permutation_dim, input_dim)
+    and outputs (batch,output_dim).
+    """
+
+    def __init__(
+        self,
+        single_trial_net: nn.Module,
+        input_dim: int,
+        output_dim: int,
+        num_fc: int = 2,
+        num_hiddens: int = 20,
+        combining_operation: str = "mean",
+    ):
+        """Permutation invariant multi-layer NN.
+            The single_trial_net is applied to each "trial" of the input
+            and is combined by the combining_operation (mean or sum).
+
+        Args:
+            single_trial_net: Network to process one trial, the combining_operation is
+                applied to its output. Taskes as input (batch, input_dim).
+                Remark: This network should be large enough as it acts on all (iid)
+                inputs seperatley and needs enough capacity to process the information
+                of all inputs.
+            input_dim: Dimensionality of input to the fully connected layers
+                (output_dimension of single_trial_net).
+            output_dim: Dimensionality of the output.
+            num_fc: Number of fully connected layer, minimum of 2.
+            num_hiddens: Number of hidden dimensions in fully-connected layers.
+            combining_operation: How to combine the permutational dimensions, one of
+                'mean' or 'sum'.
+        """
+        super().__init__()
+        self.single_trial_net = single_trial_net
+        self.input_dim = input_dim
+        self.output_dim = output_dim
+        self.num_hiddens = num_hiddens
+        self.combining_operation = combining_operation
+
+        # construct fully connected layers
+        fc_layers = [nn.Linear(input_dim, num_hiddens), nn.ReLU()]
+        for _ in range(num_fc - 2):
+            fc_layers.append(nn.Linear(num_hiddens, num_hiddens))
+            fc_layers.append(nn.ReLU())
+        fc_layers.append(nn.Linear(num_hiddens, output_dim))
+
+        self.fc_subnet = nn.Sequential(*fc_layers)
+
+    def forward(self, x):
+        """Network forward pass.
+        Args:
+            x: Input tensor (batch_size, permutation_dim, input_dim)
+        Returns:
+            Network output (batch_size, output_dim).
+        """
+        batch, permutation_dim, _ = x.shape
+
+        e = self.single_trial_net(x.view(batch * permutation_dim, -1)).view(
+            batch, permutation_dim, -1
+        )
+
+        if self.combining_operation == "mean":
+            e = e.mean(1)
+        elif self.combining_operation == "sum":
+            e = e.sum(1)
+
+        embedding = self.fc_subnet(e)
```

Comment:
I don't understand why we have another embedding net here. Shouldn't it be enough to embed each of the trials using the single_trial_net and then take the mean? 
We could then pass on the resulting mean embedding to the conditional density estimator, plus the number of trials.

Reply:
I just talked with Michael about this, and he pointed to the Deep Sets paper where the summed input is also processed in an additional fully-connected DNN. So I would stick to this structure here. 
Adding other conditions like n_trials could be done, but I would leave this for later and start here with a simple version. This would also need changes of the input shapes etc.